### PR TITLE
Fix how library handles region specific pack files

### PR DESCRIPTION
**Pull request by akshob**
Summary:
If the application has region specific languages, the library fails to fetch them.
For example, if the app has two regional Chinese languages `zh-CN` and `zh-TW`, there would be two pack files for them. `strings_zh-rCN.pack` and `strings_zh-rTW.pack` respectively. When the library tries to fetch the pack file it currently always  gets `strings_zh.pack` and ignores the region.
This change fixes this by appending the region (country) if it is initialized in the `locale`.

Test:
Added test cases for this scenario with two regional locales: `ha-NG` and `zh-TW`
Verified in the sample app that switching to region specific language, updates the text correctly

### CODE REVIEW

File: library/src/main/java/com/whatsapp/stringpacks/StringPacks.java [hunk 1]
```diff
@@ -203,7 +204,8 @@ private static boolean useSystemResources(@NonNull Locale locale) {
   }
 
   private static String getPackFileName(Locale locale) {
-    return "strings_" + locale.getLanguage();
+    return String.format("strings_%s" + (!locale.getCountry().equals("") ? "-r%s" : ""),
```
Comment (ashish29dec) [inline library/src/main/java/com/whatsapp/stringpacks/StringPacks.java:207]: For the same reason that people may want to pack multiple locales into one pack file, like `es-MX`, `es-ES`, `es` into `strings_es.pack` file via `pack_id_mapping` property, we can not hard-code this logic here.

We should again expose an API for apps to provide proper pack file name that can be appended to `strings_` based on their implementation.

```suggestion
    String packFileId = null;
    if (stringPackLocaleManager != null) {
        packFileId = stringPackLocaleManager.getPackFileIdForLocale(locale);
    }
    if (packFileId == null) {
        packFileId = locale.getLanguage();
    }
    return "strings_" + packFileId;
```

File: library/src/main/java/com/whatsapp/stringpacks/StringPacks.java [hunk 2]
```diff
@@ -192,7 +192,8 @@ private static Locale getLocaleFromContext(@NonNull Resources resources) {
    */
   private static List<String> getParentLocales(@NonNull Locale locale) {
     final ArrayList<String> parents = new ArrayList<>();
-    parents.add(locale.getLanguage());
+    parents.add(String.format("%s" + (!locale.getCountry().equals("") ? "-%s" : ""),
```
Comment (ashish29dec) [inline library/src/main/java/com/whatsapp/stringpacks/StringPacks.java:195]: `getParentLocales` method provides locale hierarchy in which StringPacks tries to find a string. Hence, it also plays an important role in fallback mechanism.
Having said that, there is a possibility that people can pack strings for `es` and `es-MX` together in `strings_es.pack` and want a fallback mechanism to be `es-MX` -> `es` -> `en` (from Android's resources).

This change won't be able to enable this fallback logic from `es-MX` to `es` and fails the purpose of packing similar strings together via `pack_id_mapping` property of `config.json`.

In my opinion we must open up an API for apps to provide their own fallback parent locales based on their config or preference.

```suggestion
    if (stringPackLocaleManager == null || stringPackLocaleManager.shouldAddLanguageAsParentForLocale(locale)) {
        parents.add(locale.getLanguage());
    }
    if (stringPackLocaleManager != null) {
        parents.add(stringPackLocaleManager.getParentLocaleForLocale(locale));
    }
```

Hence, we can expose an interface `StringPackLocaleManager` in `StringPacks.java` and a getter/setter method for the implementation of this interface. The apps can create an implementation of this interface and set their implementation from Application class.

**Review by ashish29dec — CHANGES_REQUESTED** [review]

File: library/src/main/java/com/whatsapp/stringpacks/StringPacks.java (hunk 1 above)
Comment (akshob) [reply library/src/main/java/com/whatsapp/stringpacks/StringPacks.java:207]: Yes, totally makes sense.
So, the stringPackLocaleManager should have access to the `"pack_id_mapping" : {}` that is defined in the `config.json` right? 
And if a mapping is provided for `es-MX` to `es` in `config.json` then `stringPackLocaleManager.getPackFileIdForLocale(es-MX)` would return `es` otherwise `null`.

Any suggestions on how the library will have access on the config file? Or is there a better approach?

Comment (akshob) [reply library/src/main/java/com/whatsapp/stringpacks/StringPacks.java:207]: Thoughts on this approach:

Similar to how the `assemble_string_packs.py` script generates `StringPackIds.java` we also generate `StringPacksLocaleManger.java` which would contain the `pack_id_mapping` from the `config.json`
During the application setup we would need to do something like `StringPacksLocaleManager.registerLocaleManger()`. This would give us access to the mapping locales defined by the user.

Comment (ashish29dec) [reply library/src/main/java/com/whatsapp/stringpacks/StringPacks.java:207]: @akshob I think it gives the developers more flexibility if this is not a generated file but a file created by application developers and then registered via the API.
Applications can call `StringPacks.registerStringPackLocaleManager()` inside of their Application class' `attachBaseContext()` method.
In the suggestion below I renamed the `StringPacksLocaleManager` to `StringPacksLocaleMetaDataProvider`

For example, this file can look like
```
    public class LocaleMetaDataProviderImpl implements StringPacksLocaleMetaDataProvider {
        public String getPackFileIdForLocale(Locale locale) {
            // Applications can return the appropriate pack file id for the locale, based on their mapping in `pack_id_mapping` setting
        }
        
        public boolean shouldAddLanguageAsParentForLocale(Locale locale) {
            // This is a special case because we don't want to add language as a fallback locale for some locales, for example, zh can not be a fallback locale for zh-TW even if they are packed together, because zh is written in a different script than zh-TW.
            // Applications have to handle their special cases and return true or false based on their specific cases
        }
        
        public String getParentLocaleForLocale(Locale locale) {
            // Applications can decide whether to fallback to any specific locale and not necessarily just the language
        }
    }
```

File: library/src/main/java/com/whatsapp/stringpacks/StringPacksLocaleMetaDataProvider.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package com.whatsapp.stringpacks;
+
+import java.util.Locale;
+
+public interface StringPacksLocaleMetaDataProvider {
+    /**
+     * Applications can return the appropriate pack file id for the locale,
+     * based on their mapping in `pack_id_mapping` setting
+     * @param locale
+     * @return
```
Comment (ashish29dec) [inline library/src/main/java/com/whatsapp/stringpacks/StringPacksLocaleMetaDataProvider.java:10]: I think it would be best if we can also provide an example in the method comment showing an entry in `pack_id_mapping` and then what to do in here.

for example
```suggestion
     /**
     * Applications can return the appropriate pack file id for the locale,
     * based on their mapping in `pack_id_mapping` setting.
     * This identifies which pack file should the library read from for the specific locale
     * for example,
     * // config.json
     * pack_id_mapping = {
     *   "es-rMX": "es",
     *   "es-rES": "es"
     * }
     * 
     * So, for locales "es-MX" or "es-ES" you should return "es" as the pack Id
     * 
     * @param locale
     * @return packId or null if same as locale's language
```

File: library/src/main/java/com/whatsapp/stringpacks/StringPacksLocaleMetaDataProvider.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    String getPackFileIdForLocale(Locale locale);
+
+    /**
+     * This is a special case because we don't want to add language as a fallback locale
+     * for some locales, for example, zh can not be a fallback locale for zh-TW
+     * even if they are packed together, because zh is written in a different script than zh-TW.
+     * Applications have to handle their special cases
+     * and return true or false based on their specific cases.
+     * @param locale
+     * @return
```
Comment (ashish29dec) [inline library/src/main/java/com/whatsapp/stringpacks/StringPacksLocaleMetaDataProvider.java:21]: Please explain when true is returned and when false is

File: library/src/main/java/com/whatsapp/stringpacks/StringPacksLocaleMetaDataProvider.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+     */
+    boolean shouldAddLanguageAsParentForLocale(Locale locale);
+
+    /**
+     * Applications can decide whether to fallback to any specific locale
+     * and not necessarily just the language
+     * @param locale
+     * @return
```
Comment (ashish29dec) [inline library/src/main/java/com/whatsapp/stringpacks/StringPacksLocaleMetaDataProvider.java:29]: Please explain what is returned

File: sample/app/src/test/java/com/whatsapp/stringpacks/sample/LocaleUtilTest.java [hunk 6]
```diff
@@ -42,17 +43,37 @@ public void setUp() {
     when(resources.getConfiguration()).thenReturn(configuration);
     when(resources.getAssets()).thenReturn(assetManager);
     try {
-      InputStream inputStream = getClass().getClassLoader().getResourceAsStream("strings_zh.pack");
+      InputStream inputStream = ApplicationProvider.getApplicationContext().getAssets().open("strings_zh.pack");
```
Comment (ashish29dec) [inline sample/app/src/test/java/com/whatsapp/stringpacks/sample/LocaleUtilTest.java:46]: Why this change?

File: sample/app/src/test/java/com/whatsapp/stringpacks/sample/LocaleUtilTest.java [hunk 7]
```diff
@@ -42,17 +43,37 @@ public void setUp() {
...
       when(assetManager.open("strings_zh.pack")).thenReturn(inputStream);
     } catch (IOException e) {
       Assert.fail("Test setup failure" + e);
     }
+    try {
+      InputStream inputStream = ApplicationProvider.getApplicationContext().getAssets().open("strings_zh-rTW.pack");
```
Comment (ashish29dec) [inline sample/app/src/test/java/com/whatsapp/stringpacks/sample/LocaleUtilTest.java:52]: I think you need to add this .pack file into resources directory

File: sample/app/src/test/java/com/whatsapp/stringpacks/sample/LocaleUtilTest.java [hunk 8]
```diff
@@ -42,17 +43,37 @@ public void setUp() {
...
+      when(assetManager.open("strings_zh-rTW.pack")).thenReturn(inputStream);
+    } catch (IOException e) {
+      Assert.fail("Test setup failure" + e);
+    }
+    try {
+      InputStream inputStream = ApplicationProvider.getApplicationContext().getAssets().open("strings_ha.pack");
```
Comment (ashish29dec) [inline sample/app/src/test/java/com/whatsapp/stringpacks/sample/LocaleUtilTest.java:58]: Same comment as above

**Review by ashish29dec — COMMENTED** [review]
- We should add some more unit tests in `StringPacksTest.java` to test if we are fetching correct translations for "zh-TW" and "ha-NG" as these are special cases.
- Also we need to update README.md to specify the addition of new Interface and what other developers need to do when using the library

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java [hunk 9]
```diff
@@ -32,17 +33,20 @@
...
   @Mock Configuration configuration;
   @Mock AssetManager assetManager;
+  @Mock StringPacksLocaleMetaDataProvider stringPacksLocaleMetaDataProvider;
 
   private Application application;
 
   @Before
   public void setUp() {
-    MockitoAnnotations.initMocks(this);
+    MockitoAnnotations.openMocks(this);
```
Comment (ashish29dec) [inline library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:49]: Why this change?

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java [hunk 10]
```diff
@@ -55,13 +59,24 @@ public void setUp() {
     } catch (IOException e) {
       Assert.fail("Test setup failure" + e);
     }
+    try {
+      InputStream inputStream = getClass().getClassLoader().getResourceAsStream("strings_zh-rTW.pack");
+      when(assetManager.open("strings_zh-rTW.pack")).thenReturn(inputStream);
```
Comment (ashish29dec) [inline library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:64]: Are the files in the same format as specified in [README.md](https://github.com/WhatsApp/StringPacks/blob/master/library/src/test/resources/README.md)?

Please mention the content of `zh-TW`, `ha` and `ha-NG` strings.xml as done for `zh`.
Also, what changes did you make to generate `string_ha.pack` with `ha-NG`'s translations, so that later it becomes easy for someone to add/update tests to be able to generate these files without breaking tests or re-writing them all over again

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java (hunk 9 above)
Comment (akshob) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:49]: initMocks is depricated and the replacement is openMocks

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java (hunk 10 above)
Comment (akshob) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:64]: No. These files are the pack files generated by the sample app. Strings are in `sample/app/src/main/string-packs/strings/values-ha-rNG` and `sample/app/src/main/string-packs/strings/values-zh-rTW`

Ok. I will change these to the strings that in the readme file.

File: library/src/main/java/com/whatsapp/stringpacks/StringPacksLocaleMetaDataProvider.java [hunk 11]
```diff
@@ -13,7 +13,10 @@
  * to support region specific locales.
  *
  * If all the region specific locales are mapped in `pack_id_mapping` in
```
Comment (ashish29dec) [inline library/src/main/java/com/whatsapp/stringpacks/StringPacksLocaleMetaDataProvider.java:15]: The initial part of the comment is not needed anymore. Please delete it.

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java (hunk 9 above)
Comment (ashish29dec) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:51]: This change is not a required for this PR. Please revert this one and if needed raise another PR for this change

**Review by ashish29dec — COMMENTED** [review]
My comments so far. I will review unit tests also ASAP.

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java (hunk 10 above)
Comment (akshob) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:66]: @ashish29dec I changed `string_ha.pack` and specified the format in `README.md`.

When I try to initialize the app with `zh-TW` or `ha-NG` locale `stringPacks.getString(StringPacksTestData.STRING_ID);` returns null.
The only valid string is `stringPacks.getString(1);` any other integer returns null. Not sure what I missed.

Comment (ashish29dec) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:66]: The way we generate these test `.pack` files is the same way our actual application does it.
1. Since `StringPacksTestData.STRING_ID` is 15, you must have at least ID 15 (which is `hello_world`) translation in your `app/src/main/res/values-ha-rNG/strings.xml` and `app/src/main/res/values-zh-rTW/strings.xml`
2. Also, you must have all 16 strings in your `app/src/main/res/values/strings.xml`. The ids should be same as mentioned in [README.md](https://github.com/WhatsApp/StringPacks/blob/master/library/src/test/resources/README.md#strings_zhpack). However, the values can be anything as we are not worried about English strings.
3. Have your `config.json` specify correct values to pack `ha-NG` in `ha`
4. Run the `assemble_string_packs.py` command to generate the `.pack` files.
5. Verify if the order of your generated `StringPackIds.java` is the same as specified in [README.md](https://github.com/WhatsApp/StringPacks/blob/master/library/src/test/resources/README.md#strings_zhpack)

```
{
    R.plurals.plural_list,
    R.string.hello_africa,
    R.string.hello_antartica,
    R.string.hello_arctic,
    R.string.hello_asia,
    R.string.hello_atlantic,
    R.string.hello_australia,
    R.string.hello_europe,
    R.string.hello_forests,
    R.string.hello_indian,
    R.string.hello_mountains,
    R.string.hello_north_america,
    R.string.hello_pacific,
    R.string.hello_rivers,
    R.string.hello_south_america,
    R.string.hello_world,
}
```

This should generate your correct `strings_ha.pack` and strings_zh-rTW.pack` and calling `stringPacks.getString(StringPacksTestData.STRING_ID);` should return proper values.

Comment (akshob) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:64]: Thanks for the steps! I missed looking at the StringPackIds as a reference.

I've updated the `.pack` files and added more tests.

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java [hunk 12]
```diff
@@ -70,8 +90,51 @@ public void testLanguageChange() {
...
+    stringPacks.setUp(application);
+    String zhString = stringPacks.getString(StringPacksTestData.STRING_ID);
+    assertEquals("你好，世界", zhString);
+
+    // Switch language to zh-TW
+    configuration.locale = zhTWLocale;
+    when(stringPacksLocaleMetaDataProvider.getPackFileIdForLocale(any())).thenReturn("zh-rTW");
+    when(stringPacksLocaleMetaDataProvider.shouldAddLanguageAsParentForLocale(any())).thenReturn(false);
+    when(stringPacksLocaleMetaDataProvider.getFirstChoiceLocaleInPackFileForLocale(any())).thenReturn("zh-TW");
+    when(resources.getString(anyInt())).thenReturn("世界你好！");
```
[the remark this answers is not in the document]
Comment (akshob) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java]: This is fixed in the latest change

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java [hunk 13]
```diff
@@ -70,8 +90,51 @@ public void testLanguageChange() {
...
+    stringPacks.setUp(application);
+    String zhString = stringPacks.getString(StringPacksTestData.STRING_ID);
+    assertEquals("你好，世界", zhString);
+
+    // Switch language to ha-NG
+    configuration.locale = haNGLocale;
+    when(stringPacksLocaleMetaDataProvider.getPackFileIdForLocale(any())).thenReturn(null);
+    when(stringPacksLocaleMetaDataProvider.shouldAddLanguageAsParentForLocale(any())).thenReturn(false);
+    when(stringPacksLocaleMetaDataProvider.getFirstChoiceLocaleInPackFileForLocale(any())).thenReturn("ha");
+    when(resources.getString(anyInt())).thenReturn("Sannu Duniya!");
```
[the remark this answers is not in the document]
Comment (akshob) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java]: This is fixed in the latest change

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java [hunk 14]
```diff
@@ -70,8 +90,51 @@ public void testLanguageChange() {
...
     // Switch language to en-US
     configuration.locale = enLocale;
+    when(resources.getString(anyInt())).thenReturn("Test");
     stringPacks.setUp(application);
     String enString = stringPacks.getString(StringPacksTestData.STRING_ID);
     assertEquals("Test", enString);
   }
+
+  @Test
+  public void testLanguageChangeToChineseTaiwan() {
```
[the remark this answers is not in the document]
Comment (akshob) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:138]: Makes sense

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java [hunk 15]
```diff
@@ -70,8 +90,51 @@ public void testLanguageChange() {
...
+    when(stringPacksLocaleMetaDataProvider.shouldAddLanguageAsParentForLocale(any())).thenReturn(false);
+    when(stringPacksLocaleMetaDataProvider.getFirstChoiceLocaleInPackFileForLocale(any())).thenReturn("zh-TW");
+    when(resources.getString(anyInt())).thenReturn("世界你好！");
+    stringPacks.setUp(application);
+    String zhTaiwanString = stringPacks.getString(StringPacksTestData.STRING_ID);
+    assertEquals("世界你好！", zhTaiwanString);
+  }
+
+  @Test
+  public void testLanguageChangeToHausaNigeria() {
```
[the remark this answers is not in the document]
Comment (akshob) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:159]: Makes sense

File: library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java [hunk 16]
```diff
@@ -105,74 +105,18 @@ public void testLanguageChangeFromChineseTaiwanToEnglish() {
...
-    String enString = stringPacks.getString(StringPacksTestData.STRING_ID);
-    assertEquals("Test", enString);
   }
 
   @Test
-  public void testLanguageChangeFromHausaNigeriaToEnglish() {
+  public void testSetUpApplicationInHausaNigeria() {
     configuration.locale = haNGLocale;
-    when(stringPacksLocaleMetaDataProvider.getPackFileIdForLocale(any())).thenReturn("ha");
+    when(stringPacksLocaleMetaDataProvider.getPackFileIdForLocale(any())).thenReturn(null);
```
Comment (ashish29dec) [inline library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java:113]: why are we returning null? isn't `ha-NG` being packed with `ha`.

In this PR we are handling 3 special cases and tests for these would be great: -
1. Region Specific locale packed in a separate `.pack` file (Your test for zh-TW should be enough)
2. Multiple locales packed together. Your test for ha-NG should be enough
3. Fallback logic. Since `ha` and `ha-NG` are packed together. I think we should write a test where we try to fetch a string for `ha-NG` but it only exists in `ha` and we fallback to `ha` and return that string.

Comment (akshob) [reply library/src/test/java/com/whatsapp/stringpacks/StringPacksTest.java]: The `null` was something I was testing. I have changed it back to `"ha"`
I have added a test case for the fallback logic and cleaned up comments. And also updated the README file with the contents.

@ashish29dec let me know if you have more feedback.

**Comment by ashish29dec** [comment]
Overall this looks good to me. However, I am not authorized to merge the PR. I will find someone to merge the PR.
